### ISSUE TITLE
Use recursive gateways (HTTPGatewayRouter) as a fallback when other routers fails

### ISSUE THREAD

**Issue by 2color** [ipfs/helia#665]
## Background

Right now the Routing class, which wraps the different router implementations (of which are 3) treats all routers as equal, and calls them in parallel (by using the merge from `it-merge`):

https://github.com/ipfs/helia/blob/b723a1d9910c69064eafba965d4d96d753b91297/packages/utils/src/routing.ts#L57-L60

This means that if you have both:
- `delegatedHttpRouter` (with delegated routing endpoint) 
- `HTTPGatewayRouter` (for recursive gasteways),

There is no programmatic way for your to prioritise direct retrieval over a recursive gateway.

## Suggestion

[In the following discussion](https://github.com/ipfs/service-worker-gateway/pull/391#discussion_r1813065605), the idea of using 
recursive gateways as a "last resort" fallback came up. This means that if direct retrieval fails (because there are no providers or because helia fails to retrieve directly from a provider), it falls back on the recursive gateway to fetch data.

**Comment by SgtPooki** [comment]
I don't think we want to always fallback to recursive gateways.. but a solution that could get us the desired behavior is extending the Routers interface to support a priority key and then sorting on that prior to https://github.com/ipfs/helia/blob/b723a1d9910c69064eafba965d4d96d753b91297/packages/utils/src/routing.ts#L60

**Comment by SgtPooki** [comment]
> I don't think we want to always fallback to recursive gateways.. but a solution that could get us the desired behavior is extending the Routers interface to support a priority key and then sorting on that prior to
> 
> https://github.com/ipfs/helia/blob/b723a1d9910c69064eafba965d4d96d753b91297/packages/utils/src/routing.ts#L60

I think if we want to actually delay fallback routers we might need some changes other than just a priority key, since they’re all kicking off at essentially the same time..
we could do something to indicate that certain routers should be run in parallel/serial?

**Comment by achingbrain** [comment]
You could add a delay to the `HTTPGatewayRouter` before it started it's search, that'd have the effect of allowing other routers to respond first, but... does the user care or do they just want whatever method is fastest?

**Comment by SgtPooki** [comment]
I think users will want the fastest methods, but I think the motivation here is deprioritizing the use of recursive gateways, and trying to do all the work on the user side first. 

@lidel could speak more to primary motivations

**Comment by achingbrain** [comment]
Implemented in #1084
